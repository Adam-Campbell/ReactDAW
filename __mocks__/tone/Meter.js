class Meter {
    constructor() {
        this.getLevel = jest.fn();
        this.connect = jest.fn();
        this.disconnect = jest.fn();
    }
}

export default Meter;