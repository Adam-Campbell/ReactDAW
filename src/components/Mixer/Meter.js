import React, { Component } from 'react';
import PropTypes from 'prop-types';

class Meter extends Component {
    constructor(props) {
        super(props);
        this.rafRef = null;
        this.meterElementRef = React.createRef();
    }

    componentDidMount() {
        requestAnimationFrame(this.recalculateMeter);
    }

    componentWillUnmount() {
        if (this.rafRef) {
            cancelAnimationFrame(this.rafRef);
        }
    }

    recalculateMeter = () => {
        const newLevel = window.meterNodeReferences[this.props.trackId].getLevel();
        let newPercentValue;
        const topBound = 10;
        const bottomBound = -40;
        if (newLevel >= topBound) {
            newPercentValue = 100;
        } else if (newLevel <= bottomBound) {
            newPercentValue = 0;
        } else {
            const range = 50;
            newPercentValue = (newLevel - bottomBound) / range * 100;
        }
        this.meterElementRef.current.style.height = `${newPercentValue}%`;
        this.rafRef = requestAnimationFrame(this.recalculateMeter);

    }

    render() {
        return (
            <div className={this.props.isMaster ? "mixer__master-meter-outer" : "mixer__meter-outer"}>
                <div 
                    className={this.props.isMaster ? "mixer__master-meter-inner" : "mixer__meter-inner"}
                    ref={this.meterElementRef}
                ></div>
            </div>
        );
    }
}

Meter.propTypes = {
    trackId: PropTypes.string.isRequired,
    isMaster: PropTypes.any
}

export default Meter;
