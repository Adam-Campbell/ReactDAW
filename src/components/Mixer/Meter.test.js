import React from 'react';
import { shallow } from 'enzyme';
import Meter from './Meter';

test('renders correctly', () => {
    const component = shallow(
        <Meter 
            trackId="2812513466104156"
        />
    );
    expect(component).toMatchSnapshot();
});