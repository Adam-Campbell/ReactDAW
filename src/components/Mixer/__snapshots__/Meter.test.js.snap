// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly 1`] = `
<div
  className="mixer__meter-outer"
>
  <div
    className="mixer__meter-inner"
  />
</div>
`;
